### PR TITLE
[RFC] gnrc: initial import of an asynchronous error reporting API

### PR DESCRIPTION
**Pull request by authmillenon**
This gives the gnrc stack the ability to (optionally) asynchronously report on errors. If the pseudo-module `ng_neterr` is included, the function `ng_pktbuf_release_error()` will then send a message with a given error code to every subscriber to errors on every packet snip in the packet (a thread can subscribe to errors on a packet snip via `ng_neterr_reg()`). This way an application can e.g. subscribe to the payload, a transport layer can subscribe to the transport header, a routing protocol can subscribe to its header and so on. This is an idea @haukepetersen and I had a while back and since I needed something to get my mind of the current gnrc debugging to not get crazy I started to doodle this one out. If this PR gets merged we can start replacing the `ng_pktbuf_release()`s with `ng_pktbuf_release_error()`s for the error cases throughout the network stack.

### CODE REVIEW

File: sys/include/net/ng_neterr.h [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+        msg_t msg;
+
+        msg.type = NG_NETERR_MSG_TYPE;
+        msg.content.value = err;
+
+        msg_send(&msg, pkt->err_sub);
+    }
+}
+#else
+#define ng_neterr_report(pkt, err)  (void)pkt; (void)err;
```
Comment (OlegHahm) [inline sys/include/net/ng_neterr.h]: Why not simply 
```C
#define ng_neterr_report(pkt, err)
```
?

Comment (authmillenon) [inline sys/include/net/ng_neterr.h]: I got an error (or better a warning that became an error due to `-Werror`) with that.

Comment (OlegHahm) [inline sys/include/net/ng_neterr.h]: I see, but
```C
#define ng_neterr_report(pkt, err) (void) err;
```
would be sufficient. Anyway, that's of minor importance.

**Comment by OlegHahm** [comment]
ACK

**Comment by authmillenon** [comment]
Let's postpone merging this until #3496 got merged.

**Comment by authmillenon** [comment]
Rebased and adapted for new packet buffer.

**Comment by OlegHahm** [comment]
Needs another rebase.

**Comment by authmillenon** [comment]
Rebased (please test if behavior stays the same as without this PR, there were some nasty merge conflicts)

File: sys/include/net/ng_pktbuf.h [hunk 2]
```diff
@@ -139,11 +140,24 @@ void ng_pktbuf_hold(ng_pktsnip_t *pkt, unsigned int num);
...
  * @brief   Decreases ng_pktsnip_t::users of @p pkt atomically and removes it if it
- *          reaches 0.
+ *          reaches 0 and reports a possible error through an error code, if
+ *          @ref net_ng_neterr is included.
  *
  * @param[in] pkt   A packet.
+ * @param[in] err   An error code.
  */
-void ng_pktbuf_release(ng_pktsnip_t *pkt);
+void ng_pktbuf_release_error(ng_pktsnip_t *pkt, uint32_t err);
```
Comment (haukepetersen) [inline sys/include/net/ng_pktbuf.h]: the release error is always present, even if the `neterr` module is not activated?

Comment (authmillenon) [inline sys/include/net/ng_pktbuf.h]: Yes, but read the docs.

**Comment by OlegHahm** [comment]
Needs a rebase.

**Comment by authmillenon** [comment]
Rebased and adapted to current master (and it was no fun at all)

File: sys/include/net/gnrc/neterr.h [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+#include "thread.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief   @ref core_msg type for reporting an error.
+ */
+#define GNRC_NETERR_MSG_TYPE        (0x0206)
```
Comment (OlegHahm) [inline sys/include/net/gnrc/neterr.h]: It's a general problem, but to improve the situation here, I would prefer to move this into `sys/include/net/gnrc/netapi.h`.

Comment (authmillenon) [inline sys/include/net/gnrc/neterr.h]: But this is no NETAPI message type…

File: sys/include/net/gnrc/neterr.h [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+
+/**
+ * @brief   Error code for success.
```
Comment (OlegHahm) [inline sys/include/net/gnrc/neterr.h]: Sounds a bit contradicting. ;-) Can you add an explanation what this indicates?

File: sys/include/net/gnrc/neterr.h [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+
+/**
+ * @brief   Registers the current thread for errors on a @ref gnrc_pktsnip_t.
+ *
+ * @param[in] pkt   Packet snip to register for errors.
+ *
+ * @return  0, on success.
+ * @return  EALREADY, if there already someone registered to errors on @p pkt.
+ */
+#ifdef MODULE_GNRC_NETERR
```
Comment (OlegHahm) [inline sys/include/net/gnrc/neterr.h]: Why is this necessary?

File: sys/include/net/gnrc/neterr.h [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+static inline int gnrc_neterr_reg(gnrc_pktsnip_t *pkt)
+{
+    if (pkt->err_sub != KERNEL_PID_UNDEF) {
+        return EALREADY;
+    }
+    pkt->err_sub = thread_getpid();
```
Comment (OlegHahm) [inline sys/include/net/gnrc/neterr.h]: `s/thread_getpid/sched_active_pid/`

File: sys/include/net/gnrc/neterr.h (hunk 4 above)
Comment (authmillenon) [inline sys/include/net/gnrc/neterr.h]: How? For me it makes perfect sense, honostly. It's a value in the class of error codes, that indicates success (no error).

Comment (authmillenon) [inline sys/include/net/gnrc/neterr.h]: You are the one confused so help me resolve your confusion ;-)

File: sys/include/net/gnrc/pktbuf.h [hunk 7]
```diff
@@ -141,13 +142,26 @@ void gnrc_pktbuf_hold(gnrc_pktsnip_t *pkt, unsigned int num);
...
-void gnrc_pktbuf_release(gnrc_pktsnip_t *pkt);
+void gnrc_pktbuf_release_error(gnrc_pktsnip_t *pkt, uint32_t err);
+
+/**
+ * @brief   Decreases gnrc_pktsnip_t::users of @p pkt atomically and removes it if it
+ *          reaches 0 and reports @ref GNRC_NETERR_SUCCESS.
+ *
+ * @param[in] pkt   A packet.
+ */
+static inline void gnrc_pktbuf_release(gnrc_pktsnip_t *pkt)
```
Comment (OlegHahm) [inline sys/include/net/gnrc/pktbuf.h]: Why is this needed?

File: sys/include/net/gnrc/neterr.h (hunk 3 above)
Comment (OlegHahm) [inline sys/include/net/gnrc/neterr.h]: Well, either we make this exception or add another header. Having similar message types defined in different headers are prone to create duplicates.

File: sys/include/net/gnrc/neterr.h (hunk 5 above)
Comment (authmillenon) [inline sys/include/net/gnrc/neterr.h]: So you don't have to #ifdef everytime you call this function.

File: sys/include/net/gnrc/neterr.h (hunk 6 above)
Comment (authmillenon) [inline sys/include/net/gnrc/neterr.h]: Isn't this the same?

File: sys/include/net/gnrc/neterr.h (hunk 4 above)
Comment (OlegHahm) [inline sys/include/net/gnrc/neterr.h]: For me error contradicts success. I have absolute no idea what this value should tell me. An error occurred? Or everything went fine?

File: sys/include/net/gnrc/neterr.h (hunk 3 above)
Comment (authmillenon) [inline sys/include/net/gnrc/neterr.h]: Yepp, but this is a general problem

File: sys/include/net/gnrc/neterr.h (hunk 5 above)
Comment (OlegHahm) [inline sys/include/net/gnrc/neterr.h]: Ah, sorry, overlooked the else.

File: sys/include/net/gnrc/neterr.h (hunk 6 above)
Comment (OlegHahm) [inline sys/include/net/gnrc/neterr.h]: Yes, just accessing the variable directly instead of calling a function.

File: sys/include/net/gnrc/neterr.h (hunk 3 above)
Comment (OlegHahm) [inline sys/include/net/gnrc/neterr.h]: That's what I've said.

File: sys/include/net/gnrc/pktbuf.h (hunk 7 above)
Comment (authmillenon) [inline sys/include/net/gnrc/pktbuf.h]: To not have to replace every instance of `gnrc_pktbuf_release()` in the code and to not require a user to use `gnrc_pktbuf_release_error()` when there was actually no error.

**Comment by authmillenon** [comment]
Addressed comments.

File: sys/include/net/gnrc/neterr.h (hunk 3 above)
Comment (authmillenon) [inline sys/include/net/gnrc/neterr.h]: I think I have an idea of how to solve this.

**Comment by authmillenon** [comment]
Rebased to current master.

**Comment by authmillenon** [comment]
Squashed

**Comment by haukepetersen** [comment]
ack
